fpga_interchange/arch.h

```diff
@@ -0,0 +1,1308 @@
+
+    // -------------------------------------------------
+
+    void setup_byname() const;
+
+    BelId getBelByName(IdStringList name) const;
+
+    IdStringList getBelName(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        int site_index = locInfo(bel).bel_data[bel.index].site;
+        NPNR_ASSERT(site_index >= 0);
+        const SiteInstInfoPOD &site = chip_info->sites[chip_info->tiles[bel.tile].sites[site_index]];
+        std::array<IdString, 2> ids{id(site.name.get()), IdString(locInfo(bel).bel_data[bel.index].name)};
+        return IdStringList(ids);
+    }
+
+    uint32_t getBelChecksum(BelId bel) const { return bel.index; }
+
+    void bindBel(BelId bel, CellInfo *cell, PlaceStrength strength)
+    {
+        NPNR_ASSERT(bel != BelId());
+        NPNR_ASSERT(tileStatus[bel.tile].boundcells[bel.index] == nullptr);
+
+        tileStatus[bel.tile].boundcells[bel.index] = cell;
+        cell->bel = bel;
+        cell->belStrength = strength;
+        refreshUiBel(bel);
+    }
+
+    void unbindBel(BelId bel)
+    {
+        NPNR_ASSERT(bel != BelId());
+        NPNR_ASSERT(tileStatus[bel.tile].boundcells[bel.index] != nullptr);
+        tileStatus[bel.tile].boundcells[bel.index]->bel = BelId();
+        tileStatus[bel.tile].boundcells[bel.index]->belStrength = STRENGTH_NONE;
+        tileStatus[bel.tile].boundcells[bel.index] = nullptr;
+        refreshUiBel(bel);
+    }
+
+    bool checkBelAvail(BelId bel) const { return tileStatus[bel.tile].boundcells[bel.index] == nullptr; }
+
+    CellInfo *getBoundBelCell(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        return tileStatus[bel.tile].boundcells[bel.index];
+    }
+
+    CellInfo *getConflictingBelCell(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        return tileStatus[bel.tile].boundcells[bel.index];
+    }
+
+    BelRange getBels() const
+    {
+        BelRange range;
+        range.b.cursor_tile = 0;
+        range.b.cursor_index = -1;
+        range.b.chip = chip_info;
+        ++range.b; //-1 and then ++ deals with the case of no Bels in the first tile
+        range.e.cursor_tile = chip_info->width * chip_info->height;
+        range.e.cursor_index = 0;
+        range.e.chip = chip_info;
+        return range;
+    }
+
+    Loc getBelLocation(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        Loc loc;
+        getTileXY(bel.tile, &loc.x, &loc.y);
+        loc.z = bel.index;
+        return loc;
+    }
+
+    BelId getBelByLocation(Loc loc) const;
+    BelRange getBelsByTile(int x, int y) const;
+
+    bool getBelGlobalBuf(BelId bel) const
+    {
+        // FIXME: This probably needs to be fixed!
+        return false;
+    }
+
+    bool getBelHidden(BelId bel) const { return locInfo(bel).bel_data[bel.index].category != BEL_CATEGORY_LOGIC; }
+
+    IdString getBelType(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        return IdString(locInfo(bel).bel_data[bel.index].type);
```

Comment:
something I did in nexus, which improved the code quality a bit; was to have `bel_data`, etc, functions instead of using `locInfo` and indexing, which decouples the chip data accesses a bit. If iterators where given a `Arch` pointer instead of a direct database pointer; then I think they could use these same functions in a few places too.

the implementation in nexus actually has `bel_data` resolving to a static `chip_bel_data` because iterators there have database pointers but I don't think that was actually needed, retrospectively.

https://github.com/YosysHQ/nextpnr/blob/master/nexus/arch.h#L1432
https://github.com/YosysHQ/nextpnr/blob/master/nexus/arch.h#L314

Reply:
Fixed.